merino/jobs/navigational_suggestions/domain_metadata_diff.py

```diff
@@ -0,0 +1,129 @@
+        return [entry["domain"] for entry in domain_data["domains"]]
+
+    def process_urls(self, domain_data: dict[str, list[dict[str, str]]]) -> list[str]:
+        """Process the domain list and return a list of all urls."""
+        return [entry["url"] for entry in domain_data["domains"]]
+
+    def process_categories(
+        self, domain_data: dict[str, list[dict[str, str]]]
+    ) -> list[str]:
+        """Process the domain list and return a list of all distinct categories."""
+        distinct_categories: list[str] = list(
+            set(
+                [
+                    value
+                    for entry in domain_data["domains"]
+                    for value in entry["categories"]
+                ]
+            )
+        )
+
+        return distinct_categories
+
+    def check_url_for_subdomain(
+        self, domain_data: dict[str, list[dict[str, str]]]
+    ) -> list[dict[str, str]]:
+        """Compare the domain and url to check if a subdomain occurs immediately
+        after scheme.
+        """
+        url_pattern: str = r"^https?://(www\.)?"
+        subdomain_occurences = []
+        for entry in domain_data["domains"]:
+            url_value = re.sub(url_pattern, "", entry["url"])
+            if url_value.split(".")[0] != entry["domain"]:
+                subdomain_occurences.append(
+                    {
+                        "rank": entry["rank"],
+                        "domain": entry["domain"],
+                        "url": entry["url"],
+                    }
+                )
+        return subdomain_occurences
+
+    def compare_top_picks(
+        self,
+        new_top_picks: dict[str, list[dict[str, str]]],
+        old_top_picks: dict[str, list[dict[str, str]]],
+    ) -> tuple[list[str], set[str], set[str], set[str], list[dict[str, str]]]:
+        """Compare the previous file with new data to be written in latest file."""
+        previous_data: dict[str, list[dict[str, str]]] = old_top_picks
+        new_data: dict[str, list[dict[str, str]]] = new_top_picks
+
+        categories: list[str] = self.process_categories(new_data)
+        previous_urls: list[str] = self.process_urls(previous_data)
+        new_urls: list[str] = self.process_urls(new_data)
+        previous_domains: list[str] = self.process_domains(previous_data)
+        new_domains: list[str] = self.process_domains(new_data)
+        subdomains: list[dict[str, str]] = self.check_url_for_subdomain(new_data)
+
+        unchanged_domains: set[str] = set(new_domains).intersection(
+            set(previous_domains)
+        )
+        # added_domains returns the domains not in the previous `_latest` file
+        added_domains: set[str] = set(new_domains).difference(set(previous_domains))
+        added_urls: set[str] = set(new_urls).difference(set(previous_urls))
+
+        return (categories, unchanged_domains, added_domains, added_urls, subdomains)
+
+    def create_diff_file(
```

Comment:
Nit: since it won't create a file
```suggestion
    def create_diff(
```